### DISCUSSION TITLE
Ideation:  OSS Architecture

### DISCUSSION THREAD

**Discussion by manishapriya94** [ProgramEquity/amplify#61] [Ideas]
## Challenge
To echo the larger challenge of scaling this project within this hackathon - we focus on creating a robust OSS DevSecOps workflow and infrastructure. 

**Priorities**
- [CodeSpaces](https://github.com/ProgramEquity/amplify-back-end/issues/203)
- [Actions ](https://github.com/ProgramEquity/amplify-back-end/issues/192)
- [ORM ](https://github.com/ProgramEquity/amplify-back-end/issues/61)

You might find this skillset track interesting if you hold any of the following titles: 
- Project Management
- Architecture
- Security/Cybersecurity  Analyst
- DevOps engineer
- DevSecOps 

### Problem
More than 80% of the world’s software is built on open source. Log4J and Solorigates tested organizations across the world and how robust their infrastructure was when it came to operations, discovery, and remediation. 

Maintainers navigate a very different world. Resources across tools, people, and time are afforded differently so we take on the lens of MGV when trying to adapt DevSecOps thats accessible in the projects that fuel our development heartbeats - open source. 

This is an exploration of the accessibility of securing open source as discussed at the [Open Summit](https://www.google.com/url?q=https://github.blog/2022-01-13-open-source-software-security-summit-securing-the-worlds-code-together/&sa=D&source=docs&ust=1643128107559411&usg=AOvVaw2LhkW5ITtzK4rngXaezpe3) at the White House earlier this year.  

## Goal 
We plan to use Minimum Viable Governance framework so that security across user policies, infrastructure, appsec, app structure, data, and open source collaboration can scale seamlessly with the least amount of overhead. 

MVG provides a two-tier structure for a set of open source projects. At the top level (called an “organization” on GitHub), there is a technical steering committee to make decisions about the overall direction and coordination between all the organization's projects. Underneath that top level are the individual projects, with lightweight, consensus-based governance.

### Pillars of MVG: 
**Decision Making.** MVG defaults to a consensus-based decision making process at the organization level. Consensus does not mean no objection - it means loose agreement by the stakeholders based on their dominant view, and taking outstanding objections into consideration. [This is the approach that built the net. ](https://tools.ietf.org/id/draft-resnick-on-consensus-01.html)In our experience, striving for consensus reduces friction and reduces hostile project forks. At the organization/steering committee level, decisions that can’t be made by consensus fall back to majority vote. Individual repositories are consensus-only: if a decision can’t be made by consensus, it’s appealed to the steering committee.

[Minimal Viable Risk Management](https://static1.squarespace.com/static/556340ece4b0869396f21099/t/59d66ab28fd4d2226f13aa70/1507224246417/Minimum_Viable_Information_Risk_Management_Program.pdf): This is Appsec, zero trust policy, and protected workflows. Policies within administrative workflows which limit key decision making to OSS maintainers (steering comittee). 

### Functional Specifications
**Amplify App** 
- Going from using 40 advocacy groups to 100
- Focus areas: 
  - ORM implementation
  - Database Abstraction 
  - API integrations are made through secrets 

**OSS Architecture:** 
- Supporting  200 outside collaborators that are made up of folks used to enterprise architecture with enough enablement and intuitive workflows to 
- Focus Areas: 
  - Codespaces onboarding 
  - Production & Staging
  - CI/CD optimizations with Actions
  - Vulnerability Detection and Remediation
  - Security Policy 
  - Token Scanning
  - Project board automations

## How to get started 
**Review and pick a thread**
Understand the context of the architecture 

**Pick an issue linked to discussion**
Scope the problem you're hoping to add or seek input around 

**Submitting an issue:** 
Don't see a problem discussed? Discuss within your pod and skillset team to see if it worth submitting an issue. Use this template when submitting the issue

### Table of Contents 
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-ff55ca2e-7fff-6d9c-e57a-e073789cb423"><div dir="ltr" style="margin-left:0pt;" align="left">

  | Threads | Stack
-- | -- | --
Developer Experience | Onboarding, Continuous Integration, Continuous Delivery, CodeQL Checks, Project tracking | Codespaces, Actions, Workflows, Project Boards, GitHub Advanced Security
Application Security: | Securing current actions, Ensuring secrets for all new APIsDatabase Abstraction | Secrets, Dependabot, ORM,
Operations | Set up staging environment and production environment Administrative Actions | Heroku, Actions, Repo Structure and Permission

</div></b>
### Definitions in the context of this project

**App Structure:** 
We are going from a Single Page App Structure
<img width="772" alt="Screen Shot 2022-01-26 at 1 31 37 PM" src="https://user-images.githubusercontent.com/9143339/151250341-0b42b1fa-045b-44ec-933d-f58860ef0811.png">

To a CRUD App Structure 
<img width="739" alt="Screen Shot 2022-01-26 at 1 31 28 PM" src="https://user-images.githubusercontent.com/9143339/151250346-a1d9b6a4-e635-44a7-ba6e-36d186bbc545.png">

**OSS Architecture:** 
OSS Architecture refers to all of the tools, workflows, and developer contribution pathways we're condensing under the DevSecOps model that works for open source - GitHub flows within an MVG framework. 

<img width="1327" alt="Screen Shot 2022-01-26 at 1 08 59 PM" src="https://user-images.githubusercontent.com/9143339/151247329-f928d207-25dd-4cef-aafb-8ff6f123b7c1.png">

**Developer Experience**
- Developer experience speaks to all of the procedures of contribution flow onboarding, scoping project or issue, writing and testing code, code review, and the CI that enables their code to be merged. 
- MVG Framework Consideration: The more important matter is to integrate practices and embed them deeply into a process flow that crosses data and information silos from secure development through deployment and then to operations. To this end, continuous integration builds on test automation to ensure that as new code is added to an application, continuous integration focuses on quality at the application level as new or modified code is integrated. 
- Our focus of this hackathon is to ensure onboarding takes minimal effort while making sure we check code per merge from an administrative, project, and security perspective
- Current Contribution Flow 

<img width="619" alt="Screen Shot 2022-01-26 at 1 31 08 PM" src="https://user-images.githubusercontent.com/9143339/151250247-32d80836-0781-46ca-8b95-abe3bdcf7061.png">

- Threads: 
  - [Changing Contributor Onboarding: Codespaces](https://github.com/ProgramEquity/amplify-back-end/discussions/170#discussioncomment-2010369)

AppSec: 
**Background:** 
AppSec tools like [SAST (Static Application Security Testing)](https://www.whitesourcesoftware.com/resources/blog/sast-static-application-security-testing/), [DAST (Dynamic Application Security Testing)](https://www.whitesourcesoftware.com/resources/blog/dast-dynamic-application-security-testing/), and others that address issues in proprietary software have become staples of the developer’s security toolbox. In addition, an AppSec strategy also needs to detect open source components with known vulnerabilities, and that’s where [SCA (Software Composition Analysis)](https://www.whitesourcesoftware.com/resources/blog/software-composition-analysis/) tools play an important role. Both SAST and SCA tools address vulnerabilities management. Each one operates differently, covers a different set of vulnerabilities, and even integrates into different stages of the software development lifecycle (SDLC).

**Secure Workflows minimize risk across supply chain**
- Code scanning with CodeQL - Search for potential security vulnerabilities and coding errors in your code. For more information, see "[About code scanning."](https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/about-code-scanning)
- Secret scanning - Detect secrets, for example keys and tokens, that have been checked into the repository from API integrations. For more information, see "[About secret scanning](https://docs.github.com/en/github/administering-a-repository/about-secret-scanning)."
- Dependency review through Dependabot- Show the full impact of changes to dependencies and see details of any vulnerable versions before you merge a pull request. For more information, see "[About dependency review.](https://docs.github.com/en/code-security/supply-chain-security/about-dependency-review)"
- Report vulnerabilities through Security Lab. Maintainers figure out the ‘how’ with their security policy 

Operations: 
Operations refers to administrative practices for staging and monitoring production as well as policies and permissions.

**Comment by manishapriya94** [comment]
Category: Developer Experience - Onboarding with [Codespaces ](https://github.com/features/codespaces)

**Use case:** speed up onboarding (similar to dev containers) in a cloud environment for optimal onboarding experience for both hackathon participants and student mentorship programs

**Context:** We currently use the contributors guide for people to set up local environments in Mac, Linux, and Windows. It can sometimes create barriers for both students as well as those who would like to start contributing immediately. 

_Mac users get the following error:_
<img width="796" alt="Screen Shot 2022-01-20 at 2 50 20 PM" src="https://user-images.githubusercontent.com/9143339/150434670-96faf376-e5f9-4d93-bb61-e0d5a1d010a7.png">

@JamesMGreene has started the following issue to implement within backend:

**Reply by manishapriya94** [reply]
Setting up CI in frontend

Use Case: this will allow us to create tests for frontend instead of just backend

**Comment by manishapriya94** [comment]
Category: Operations - Production Environments for Staging and Production 

Use case: have a different environment for hackathon and one that advocacy groups continue to use

https://github.com/ProgramEquity/amplify-front-end/issues/80

**Comment by manishapriya94** [comment]
Category: AppSec - Security Policy
<img width="804" alt="Screen Shot 2022-02-02 at 1 13 57 PM" src="https://user-images.githubusercontent.com/9143339/152238424-6f2809d0-f964-4bdc-97b2-385914128bf7.png">

**Comment by JarLob** [comment]
Hi, I have looked the workflows in the repository from security perspective. Good practices are in place, like permissions and hash pinning. No injections or `pull_request_target` usage. The only thing to keep in mind secrets in `integration-tests.yaml` are available only to pull requests from the same repo. For pull requests from forks the workflow will fail. That's the secure behavior. If you want to run tests for pull requests from forks a manual trigger (label or comment from a maintainer) can be added after the changes are reviewed. As a side note dispatch event will have the access to the secrets too, but I assume everything merged has been reviewed, so it is safe.

**Comment by manishapriya94** [comment]
Dev Experience: Adding bugs to your workflows
https://github.co/3y2jfAb
